lib/features/build_server/compiler_interface.dart

```diff
@@ -0,0 +1,270 @@
+import 'dart:io';
+import 'dart:isolate';
+import 'dart:typed_data';
+
+import 'package:flyde/core/async/connect.dart';
+import 'package:flyde/core/async/interface.dart';
+import 'package:flyde/core/fs/configs/compiler_config.dart';
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:flyde/core/networking/protocol/compile_status.dart';
+import 'package:flyde/features/build_server/cache/project_cache.dart';
+import 'package:meta/meta.dart';
+
+import 'compiler.dart';
+
+/// List of constant message ids to be used in inter-isolate communication.
+class _MessageIdentifiers {
+  static const init = 'init';
+  static const build = 'build';
+  static const update = 'update';
+  static const sync = 'sync';
+  static const stateUpdate = 'stateUpdate';
+  static const hasCapacity = 'hasCapacity';
+  static const getBinary = 'getBinary';
+}
+
+/// Interface for the compiler running in a seperate [Isolate].
+@visibleForTesting
+class WorkerInterface extends Interface with CompilerStatusDelegate {
+  // ignore: unused_field
+  @visibleForTesting
+  static WorkerInterface? instance;
+
+  /// Flag to store if the worker needs to be ininitalized.
+  bool _requiresInit = true;
+
+  /// `true` if the worker can accept a new build request.
+  bool _hasCapacity = true;
+
+  /// The internal `Compiler` instance.
+  late Compiler _compiler;
+
+  @visibleForTesting
+  WorkerInterface(SpawnedIsolate isolate) : super(isolate);
+
+  /// Starts a new worker or throws an error if already running.
+  @visibleForTesting
+  static void start(SendPort sendPort, ReceivePort receivePort) {
+    if (instance != null) {
+      throw StateError('A worker instance is already running in this isolate.');
+    }
+
+    instance = WorkerInterface(
+      SpawnedIsolate(Isolate.current, receivePort)..sendPort = sendPort,
+    )..ready.complete();
+  }
+
+  @visibleForTesting
+  @override
+  Future<void> onMessage(InterfaceMessage message) async {
+    //* Respond to init requests.
+    if (message.name == _MessageIdentifiers.init &&
+        _requiresInit &&
+        message.args is List<dynamic>) {
+      final args = message.args as List<dynamic>;
+      final files = args[0] as Map<String, String>;
+      final config = args[1] as CompilerConfig;
+      final cache = args[2] as ProjectCache;
+
+      _compiler = Compiler(config, files, cache);
+      _compiler.delegate = this;
+      _requiresInit = false;
+      message.respond(isolate.sendPort, null);
+    }
+
+    //! Do not answer requests if the worker is not initialized.
+    if (_requiresInit) {
+      return;
+    }
+
+    //* Respond to capacity requests.
+    if (message.name == _MessageIdentifiers.hasCapacity) {
+      message.respond(isolate.sendPort, _hasCapacity);
+    }
+
+    //! Do not handle build requests if the worker is bussy.
+    if (!_hasCapacity) {
+      return;
```

Comment:
Maybe respond with some message "BUSY"? Especially for the `build` message, as that one would not timeout as with `expectResponse`

Reply:
Maybe a `message.respondError(error)` function could be useful for this. But that would require a new layer of error handling and should be somehow be streamlined with the web socket communication. That would require a some error handling mechanisms in event_synchronizer. 

Long story short: Good idea, much work. I'd prefer to address this in a later PR

Reply:
Ok. But please ensure some sort of timeout for `build`, maybe using the progressUpdates? If the compilation is not started in x seconds -> timeout?

Reply:
I'll think about something we can do about it 👍